File: 002_stack_next_front_node_back_mongo/002_05-referencias-etapa-criacao-estrutura/snippets-backend/repository.example.ts
// repository.example.ts
// Camada de acesso a dados

import { Product, IProduct } from "../models/Product";

export interface CreateProductData {
  name: string;
  description: string;
  price: number;
}

export interface UpdateProductData {
  name?: string;
  description?: string;
  price?: number;
  isActive?: boolean;
}

export class ProductRepository {
  async findAll(): Promise<IProduct[]> {
    return Product.find({ isActive: true }).sort({ createdAt: -1 });
  }

  async findById(id: string): Promise<IProduct | null> {
    return Product.findById(id);
  }

  async findByName(name: string): Promise<IProduct | null> {
    return Product.findOne({ name, isActive: true });
  }

  async create(data: CreateProductData): Promise<IProduct> {
    const product = new Product(data);
    return product.save();
  }

  async update(id: string, data: UpdateProductData): Promise<IProduct | null> {
    return Product.findByIdAndUpdate(
      id,
      { $set: data },
      { new: true, runValidators: true }
    );
  }

  async delete(id: string): Promise<boolean> {
    // Soft delete
    const result = await Product.findByIdAndUpdate(
      id,
      { $set: { isActive: false } },
      { new: true }
    );
    return !!result;
  }

  async hardDelete(id: string): Promise<boolean> {
    // Hard delete (use com cuidado)
    const result = await Product.findByIdAndDelete(id);
    return !!result;
  }

  async count(): Promise<number> {
    return Product.countDocuments({ isActive: true });
  }
}

// Instância única
export const productRepository = new ProductRepository();

File: 003_stack_next_front_python_back_mongo/003_05-referencias-etapa-criacao-estrutura/snippets-backend/repository.example.py
# repository.example.py
# Camada de acesso a dados (async)

from motor.motor_asyncio import AsyncIOMotorDatabase
from bson import ObjectId
from typing import List, Optional
from datetime import datetime

class ProductRepository:
    def __init__(self, db: AsyncIOMotorDatabase):
        self.collection = db.products
    
    async def find_all(self) -> List[dict]:
        """Busca todos os produtos ativos"""
        cursor = self.collection.find({"is_active": True})
        products = await cursor.to_list(length=100)
        return products
    
    async def find_by_id(self, product_id: str) -> Optional[dict]:
        """Busca produto por ID"""
        try:
            product = await self.collection.find_one({"_id": ObjectId(product_id)})
            return product
        except Exception:
            return None
    
    async def find_by_name(self, name: str) -> Optional[dict]:
        """Busca produto por nome"""
        product = await self.collection.find_one({
            "name": name,
            "is_active": True
        })
        return product
    
    async def create(self, product_data: dict) -> dict:
        """Cria novo produto"""
        product_data["created_at"] = datetime.utcnow()
        product_data["updated_at"] = datetime.utcnow()
        product_data["is_active"] = True
        
        result = await self.collection.insert_one(product_data)
        product_data["_id"] = result.inserted_id
        return product_data
    
    async def update(self, product_id: str, update_data: dict) -> Optional[dict]:
        """Atualiza produto"""
        update_data["updated_at"] = datetime.utcnow()
        
        try:
            result = await self.collection.find_one_and_update(
                {"_id": ObjectId(product_id)},
                {"$set": update_data},
                return_document=True
            )
            return result
        except Exception:
            return None
    
    async def delete(self, product_id: str) -> bool:
        """Soft delete (marca como inativo)"""
        try:
            result = await self.collection.update_one(
                {"_id": ObjectId(product_id)},
                {"$set": {"is_active": False, "updated_at": datetime.utcnow()}}
            )
            return result.modified_count > 0
        except Exception:
            return False
    
    async def count(self) -> int:
        """Conta produtos ativos"""
        return await self.collection.count_documents({"is_active": True})
